src/test_parser.rs

```diff
@@ -925,3 +932,220 @@ fn parse_display_string_errors() {
+fn complex_item_visitor() {
+    let mut holder = Holder::default();
+    let parser = Parser::new("12;x=7;y=-5");
+    let visit = HolderVisitor {
+        holder: &mut holder,
+    };
+    parser
+        .parse_item_with_visitor(visit)
+        .expect("successful parse");
+    assert_eq!(holder.point, (7, -5));
+}
+
+#[test]
+fn complex_list_visitor() {
+    #[derive(Default, Debug)]
+    struct ListHolder {
+        list: Vec<Holder>,
+        point: Point,
+    }
+
+    struct OuterListVisitor<'a> {
+        list: &'a mut Vec<ListHolder>,
+    }
+
+    impl ListVisitor<'_> for OuterListVisitor<'_> {
+        type Error = Infallible;
+        fn entry<'ev>(&mut self) -> Result<impl EntryVisitor<'ev>, Self::Error> {
+            self.list.push(ListHolder::default());
+            let list = self.list.last_mut().unwrap();
+            Ok(HolderListVisitor { list })
+        }
+    }
+
+    struct HolderListVisitor<'a> {
+        list: &'a mut ListHolder,
+    }
+
+    impl<'input> ItemVisitor<'input> for HolderListVisitor<'_> {
+        type Error = Infallible;
+        fn bare_item<'pv>(
+            self,
+            _bare_item: BareItemFromInput<'input>,
+        ) -> Result<impl ParameterVisitor<'pv>, Self::Error> {
+            panic!("expecting an inner list");
+            #[allow(unreachable_code)] // We need a concrete type here.
+            Ok(Ignored)
+        }
+    }
+
+    impl EntryVisitor<'_> for HolderListVisitor<'_> {
+        fn inner_list<'ilv>(self) -> Result<impl InnerListVisitor<'ilv>, Self::Error> {
+            Ok(self)
+        }
+    }
+
+    impl InnerListVisitor<'_> for HolderListVisitor<'_> {
+        type Error = Infallible;
+
+        fn item<'iv>(&mut self) -> Result<impl ItemVisitor<'iv>, Self::Error> {
+            self.list.list.push(Holder::default());
+            let holder = self.list.list.last_mut().unwrap();
+            Ok(HolderVisitor { holder })
+        }
+
+        fn finish<'pv>(self) -> Result<impl ParameterVisitor<'pv>, Self::Error> {
+            let point = &mut self.list.point;
+            Ok(PointParameterHolder { point })
+        }
+    }
+
+    let mut list = Vec::default();
+    let parser = Parser::new("(1;x=4 2;y=5 3);x=1;y=2,(4;x=12;y=33), ()");
+    let mut visit = OuterListVisitor { list: &mut list };
+    parser
+        .parse_list_with_visitor(&mut visit)
+        .expect("successful parse");
+
+    assert_eq!(list.len(), 3);
+    assert_eq!(list[0].list.len(), 3);
+    assert_eq!(list[0].list[0].v, 1);
+    assert_eq!(list[0].list[0].point, (4, 0));
+    assert_eq!(list[0].list[1].v, 2);
+    assert_eq!(list[0].list[1].point, (0, 5));
+    assert_eq!(list[0].list[2].v, 3);
+    assert_eq!(list[0].list[2].point, (0, 0));
+    assert_eq!(list[0].point, (1, 2));
+
+    assert_eq!(list[1].list.len(), 1);
+    assert_eq!(list[1].list[0].v, 4);
+    assert_eq!(list[1].list[0].point, (12, 33));
+    assert_eq!(list[1].point, (0, 0));
+
+    assert!(list[2].list.is_empty());
```

Comment:
Is there a reason to check all of these individually, instead of `assert_eq!(list, [...])`? As far as I can tell, the current format just allows us to write `(4, 0)` instead of `Point { x: 4, y: 0 }`, which doesn't seem like that much of a conciseness improvement given repeated indexing and field accesses.

Reply:
Only because I didn't want to implement `PartialEq` and do a full declaration of the type, but I can try it out.